### PR TITLE
Reworking lambda executor to use a single container API / break up container and lambda logic

### PR DESCRIPTION
**Pull request by dfangl**
This PR refactors the lambda executor logic to untie the lambda executor and container logic.
The goal is, to provide a interface for starting and interacting with docker (or maybe even other container runtimes someday), which the lambda executor then uses.

- [x] Add container interface
- [x] implement container interface with docker subprocess
- [x] Refactor lambda executor to make use of it
- [x] Make docker - reuse work
- [x] Remove leftover code
- [x] Add new tests and remove/adapt the ones depending hard on the current implementation
- [x] Adapt Localstack-Ext
- [x] Better documentation of the container interface

### CODE REVIEW

**Comment by thrau** [comment]
@dfangl i think if we skip the "better documentation" for now, we can start doing the localstack-ext integration ;-)

File: localstack/services/awslambda/lambda_executors.py [hunk 1]
```diff
@@ -876,87 +733,127 @@ def prepare_event(self, environment, event_body):
...
+            result = to_str(result).strip()
+        except Exception:
+            pass
+        log_output = to_str(log_output).strip()
+        return_code = process.returncode
+
+        # Note: The user's code may have been logging to stderr, in which case the logs
+        # will be part of the "result" variable here. Hence, make sure that we extract
+        # only the *last* line of "result" and consider anything above that as log output.
+        # TODO: not sure if this code is needed/used
```
Comment (whummer) [inline localstack/services/awslambda/lambda_executors.py:823]: Good catch regarding the `TODO` comment. This logic should only be required for the Docker-based Lambda executors (as lambci/docker-lambda is using stdout/stderr for different purposes). I think lines 820-832 can probably be removed.

File: localstack/utils/docker.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+from localstack.utils.bootstrap import PortMappings
+from localstack.utils.common import TMP_FILES, rm_rf, safe_run, save_file, short_uid
+
+LOG = logging.getLogger(__name__)
+
+
+@unique
+class DockerContainerStatus(Enum):
+    DOWN = -1
+    NOT_EXISTENT = 0
```
Comment (whummer) [inline localstack/utils/docker.py:18]: nitpick: I'd probably rename this to either `NON_EXISTENT` or `NOT_CREATED`

File: localstack/utils/docker.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+
+class NoSuchContainer(ContainerException):
+    def __init__(self, message, container_name_or_id, stdout=None, stderr=None) -> None:
+        super().__init__(message, stdout, stderr)
+        self.container_name_or_id = container_name_or_id
+
+
+class NoSuchImage(ContainerException):
+    def __init__(self, message, image_name, stdout=None, stderr=None) -> None:
```
Comment (whummer) [inline localstack/utils/docker.py:36]: nit: we could probably make the `message` argument optional here:
```
     def __init__(self, image_name, message=None, stdout=None, stderr=None):
         message = message or f'Docker image {image_name} not found'
         super().__init__(message, stdout, stderr)
```

Also, not sure if the type hint `-> None` adds much value for the constructors here?

File: localstack/utils/docker.py [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+            '--format="{{ .Config.Entrypoint }}"',
+            docker_image,
+        ]
+
+        LOG.debug(cmd)
+        try:
+            run_result = safe_run(cmd)
+        except subprocess.CalledProcessError as e:
+            if "No such image" in e.stdout.decode(config.DEFAULT_ENCODING):
+                raise NoSuchImage("Image not found", docker_image, e.stdout, e.stderr)
```
Comment (whummer) [inline localstack/utils/docker.py:187]: See comment above - `message` argument ("Image not found") seems a bit redundant here.

File: localstack/utils/docker.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+            else:
+                return stdout, stderr
+        except subprocess.CalledProcessError as e:
+            if "Unable to find image" in e.stderr.decode(config.DEFAULT_ENCODING):
+                raise NoSuchImage("Image not found", image_name or "", e.stdout, e.stderr)
+            if "No such container" in e.stderr.decode(config.DEFAULT_ENCODING):
+                raise NoSuchContainer(
+                    "Docker container not found", container_name, e.stdout, e.stderr
+                )
+            else:
```
Comment (whummer) [inline localstack/utils/docker.py:303]: nit: `else` technically not required here

File: tests/integration/docker/test_docker.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    yield _create_container
+
+    for c in containers:
+        try:
+            docker_client.remove_container(c)
+        except Exception:
+            LOG.warning("failed to remove test container %s", c)
+
+
+class TestDockerClient:
```
Comment (whummer) [inline tests/integration/docker/test_docker.py:65]: Out of curiosity: I'm assuming we're not extending from `unittest.TestCase` here, to be able to use the `pytest.fixture`s in the test? (or could we combine both?)

**Review by whummer — COMMENTED** [review]
Absolutely fantastic PR @dfangl , great job. 🎉 Can't express how excited I am to get this merged soon! :)

File: localstack/services/awslambda/lambda_executors.py (hunk 1 above)
Comment (thrau) [reply localstack/services/awslambda/lambda_executors.py:823]: it's also needed for the local node runners which i modified to also output the json they return: https://github.com/localstack/localstack/pull/4267/commits/6dd15bd67ddc78a2409c05f40e23ded7ff25c7a2#diff-c01a1cd0b361abb7ab2e059723a7c44b41b32a4299f4309b61cf5616f39993e3R953

File: tests/integration/docker/test_docker.py (hunk 6 above)
Comment (thrau) [reply tests/integration/docker/test_docker.py:65]: the pytest fixtures only work with unittest TestCase classes with silly workarounds/annotations. there's no value in adding `unittest.TestCase` as a base class in my opinion.

Comment (thrau) [reply tests/integration/docker/test_docker.py:65]: technically even the class is unnecessary, but it groups the tests nicely :-)

File: localstack/services/awslambda/lambda_executors.py (hunk 1 above)
Comment (whummer) [reply localstack/services/awslambda/lambda_executors.py:823]: 👍 Thanks for clarifying.

File: localstack/utils/docker.py (hunk 3 above)
Comment (dfangl) [reply localstack/utils/docker.py:36]: Value, no, it was actually autogenerated by VS Code when creating the constructor.

**Review by whummer — APPROVED** [review]
Looks awesome, thanks for pushing this huge change over the line @dfangl ! 🚀

**Pull request merged**